src/IO/WriteBufferFromS3.cpp

```diff
@@ -122,7 +123,7 @@ void WriteBufferFromS3::nextImpl()
 
 void WriteBufferFromS3::allocateBuffer()
 {
-    if (total_parts_uploaded != 0 && total_parts_uploaded % request_settings.upload_part_size_multiply_parts_count_threshold == 0)
```

Comment:
Here was the race. `total_parts_uploaded` was the number of completely uploaded parts, not just requested for upload. So when it was uploading a lot of data `total_parts_uploaded` and `upload_part_size` were updated with a delay which could lead to that error `part number is too big`.